### PR TITLE
implement global config client for tikv

### LINKED ISSUE #11686: Implement Global Config
## Development Task

[design doc](https://docs.google.com/document/d/17ccPaIq_N6GXOMHRZmcsM5mAIiWGyR8ntiZUbYs3hH0/edit?n=Global_variables_rule_them_ALL_Design_Doc#)

implement tikv pd client

### PR DESCRIPTION
**Pull request by LemonHX**
Signed-off-by: lemonhx <lemonhx@lemonhx.tech>

### What is changed and how it works?

Close #11686

What's Changed:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with None.
If this PR will be picked to release branch, then a release note is probably required.
```

### CODE REVIEW

File: components/pd_client/src/errors.rs [hunk 1]
```diff
@@ -38,6 +40,8 @@ impl ErrorCodeExt for Error {
             Error::RegionNotFound(_) => error_code::pd::REGION_NOT_FOUND,
             Error::StoreTombstone(_) => error_code::pd::STORE_TOMBSTONE,
             Error::Other(_) => error_code::pd::UNKNOWN,
+            // todo: what? why isn't there
```
Comment (sticnarf) [inline components/pd_client/src/errors.rs:43]: What does this todo mean?

File: components/error_code/src/pd.rs [hunk 2]
```diff
@@ -10,5 +10,6 @@ define_error_codes!(
     GRPC => ("gRPC", "", ""),
     REGION_NOT_FOUND => ("RegionNotFound", "", ""),
     STORE_TOMBSTONE => ("StoreTombstone", "", ""),
-    UNKNOWN => ("Unknown", "", "")
+    UNKNOWN => ("Unknown", "", ""),
+    GLOBAL_CONFIG_NOT_FOUNT => ("GlobalConfigNotFOund","","")
```
Comment (sticnarf) [inline components/error_code/src/pd.rs:14]: Prefer to keep `UNKNOWN` the last one.

File: components/pd_client/src/util.rs [hunk 3]
```diff
@@ -775,5 +775,8 @@ pub fn check_resp_header(header: &ResponseHeader) -> Result<()> {
         ErrorType::RegionNotFound => Err(Error::RegionNotFound(vec![])),
         ErrorType::Unknown => Err(box_err!(err.get_message())),
         ErrorType::Ok => Ok(()),
+        ErrorType::GlobalConfigNotFound => {
```
Comment (sticnarf) [inline components/pd_client/src/util.rs:778]: Prefer to put this before `Ok`

File: components/pd_client/src/errors.rs (hunk 1 above)
Comment (LemonHX) [reply components/pd_client/src/errors.rs:43]: oh resolved already

File: components/error_code/src/pd.rs (hunk 2 above)
Comment (LemonHX) [reply components/error_code/src/pd.rs:14]: OK

File: components/pd_client/src/util.rs (hunk 3 above)
Comment (LemonHX) [reply components/pd_client/src/util.rs:778]: OK

**Review by sticnarf — COMMENTED** [review]
Not sure about where this will be used, but the code itself looks not bad.

I will approve when you make your full feature design public.

**Comment by sticnarf** [comment]
Please fix rustfmt.

**Review by sticnarf — APPROVED** [review]
LGTM. But I notice the interfaces are all sync. Please take care when using them, so there will be no critical tasks be interfered and will not cause much performance trouble.

File: components/error_code/src/pd.rs [hunk 4]
```diff
@@ -10,5 +10,6 @@ define_error_codes!(
     GRPC => ("gRPC", "", ""),
     REGION_NOT_FOUND => ("RegionNotFound", "", ""),
     STORE_TOMBSTONE => ("StoreTombstone", "", ""),
+    GLOBAL_CONFIG_NOT_FOUNT => ("GlobalConfigNotFound","",""),
```
Comment (sticnarf) [inline components/error_code/src/pd.rs:13]: Oops, just find a typo here.

File: components/pd_client/src/client.rs [hunk 5]
```diff
@@ -282,6 +283,62 @@ impl fmt::Debug for RpcClient {
 const LEADER_CHANGE_RETRY: usize = 10;
 
 impl PdClient for RpcClient {
+    fn load_global_config(&self, list: Vec<String>) -> Result<HashMap<String, String>> {
+        use kvproto::pdpb::LoadGlobalConfigRequest;
+        let mut req = LoadGlobalConfigRequest::new();
+        for elem in list {
+            req.names.push(elem);
```
Comment (yuqi1129) [inline components/pd_client/src/client.rs:290]: maybe we can directly use `append` to add the whole vec

File: components/pd_client/src/client.rs [hunk 6]
```diff
@@ -282,6 +283,62 @@ impl fmt::Debug for RpcClient {
...
+        let res = sync_request(&self.pd_client, LEADER_CHANGE_RETRY, |client| {
+            client.store_global_config(&req)
+        })?;
+        if res.has_error() {
+            error!(
+                "error received in store global config {:?}",
+                res.get_error()
+            );
+            Err(box_err!("{:?}", res.get_error()))
+        } else {
```
Comment (yuqi1129) [inline components/pd_client/src/client.rs]: unnecessary `else`

File: components/pd_client/src/lib.rs [hunk 7]
```diff
@@ -75,6 +77,21 @@ pub const INVALID_ID: u64 = 0;
 /// creating the PdClient is enough and the PdClient will use this cluster id
 /// all the time.
 pub trait PdClient: Send + Sync {
+    /// Load a list of GlobalConfig
+    fn load_global_config(&self, _list: Vec<String>) -> PdFuture<HashMap<String, String>> {
+        unimplemented!();
+    }
+
+    /// Store a list of GlobalConfig
+    fn store_global_config(&self, _list: HashMap<String, String>) -> PdFuture<()> {
```
Comment (SunRunAway) [inline components/pd_client/src/lib.rs:86]: Why need store?
Only TiDB needs to modify a global variable

Comment (LemonHX) [reply components/pd_client/src/lib.rs:86]: but we need to implement all methods

File: components/pd_client/src/client.rs (hunk 5 above)
Comment (LemonHX) [reply components/pd_client/src/client.rs:290]: doesn't exist append method

Comment (sticnarf) [reply components/pd_client/src/client.rs:290]: The whole loop can be replaced with `req.set_names(list.into());`

File: components/pd_client/src/lib.rs (hunk 7 above)
Comment (SunRunAway) [reply components/pd_client/src/lib.rs:86]: We definitely do not need to define and/or implement "store_global_config" method in TiKV

cc @crazycs520

File: components/pd_client/src/client.rs [hunk 8]
```diff
@@ -282,6 +283,109 @@ impl fmt::Debug for RpcClient {
...
+        &self,
+    ) -> PdFuture<grpcio::ClientSStreamReceiver<pdpb::WatchGlobalConfigResponse>> {
+        use kvproto::pdpb::WatchGlobalConfigRequest;
+        let req = WatchGlobalConfigRequest::default();
+        let executor = |client: &Client, req| match client
+            .inner
+            .rl()
+            .client_stub
+            .clone()
+            .watch_global_config(&req)
```
Comment (sticnarf) [inline components/pd_client/src/client.rs:379]: This method does not block. The outer `watch_global_config` needn't return a Future.

File: components/pd_client/src/lib.rs (hunk 7 above)
Comment (JmPotato) [reply components/pd_client/src/lib.rs:86]: If it's not needed, we can use `unimplemented!()`.

Comment (sticnarf) [reply components/pd_client/src/lib.rs:86]: If it's really not needed, it should be removed from the rpc proto...

Comment (crazycs520) [reply components/pd_client/src/lib.rs:86]: TiKV PD client currently no need to implement `store_global_config`.

File: components/pd_client/src/client.rs [hunk 9]
```diff
@@ -282,6 +283,55 @@ impl fmt::Debug for RpcClient {
...
+            Err(err) => Box::pin(async move { Err(box_err!("{:?}", err)) }) as PdFuture<_>,
+        };
+        self.pd_client
+            .request(req, executor, LEADER_CHANGE_RETRY)
+            .execute()
+    }
+
+    fn watch_global_config(
+        &self,
+    ) -> Result<grpcio::ClientSStreamReceiver<pdpb::WatchGlobalConfigResponse>> {
```
Comment (zhongzc) [inline components/pd_client/src/client.rs:327]: Are there any tests for these implementations?

I'm concerned about `ClientSStreamReceiver` can work well out of grpcio runtime.

File: tests/failpoints/cases/test_pd_client.rs [hunk 10]
```diff
@@ -102,6 +104,39 @@ fn test_pd_client_deadlock() {
...
+    let (mut server, client) = new_test_server_and_client(ReadableDuration::millis(100));
+    let client = Arc::new(client);
+    use futures::StreamExt;
+    let j = std::thread::spawn(move || {
+        let _ = futures::executor::block_on(async move {
+            let mut r = client.watch_global_config().unwrap();
+            while let Some(r) = r.next().await {
+                match r {
+                    Ok(i) => {
+                        println! {"received: {:?}",i};
```
Comment (zhongzc) [inline tests/failpoints/cases/test_pd_client.rs:118]: please remove println and use assert instead.

File: components/test_pd/src/server.rs [hunk 11]
```diff
@@ -176,6 +177,43 @@ impl<C: PdMocker> Clone for PdMock<C> {
 }
 
 impl<C: PdMocker + Send + Sync + 'static> Pd for PdMock<C> {
+    fn load_global_config(
+        &mut self,
+        ctx: RpcContext<'_>,
+        req: LoadGlobalConfigRequest,
+        sink: UnarySink<LoadGlobalConfigResponse>,
+    ) {
+        hijack_unary(self, ctx, sink, |c| c.load_global_config(&req))
```
Comment (zhongzc) [inline components/test_pd/src/server.rs:186]: how about testing load_global_config also?

File: components/pd_client/src/client.rs [hunk 12]
```diff
@@ -282,6 +283,55 @@ impl fmt::Debug for RpcClient {
...
+        let executor = |client: &Client, req| match client
+            .inner
+            .rl()
+            .client_stub
+            .clone()
+            .load_global_config_async(&req)
+        {
+            Ok(grpc_response) => Box::pin(async move {
+                {
+                    let mut res = HashMap::new();
```
Comment (zhongzc) [inline components/pd_client/src/client.rs:299]: move into `Ok` branch.

Comment (zhongzc) [reply components/pd_client/src/client.rs:299]: Then you can construct a hashmap with `Hashmap::with_capacity`.

File: components/test_pd/src/server.rs [hunk 13]
```diff
@@ -176,6 +177,43 @@ impl<C: PdMocker> Clone for PdMock<C> {
...
+        &mut self,
+        ctx: RpcContext<'_>,
+        _req: WatchGlobalConfigRequest,
+        mut sink: ServerStreamingSink<WatchGlobalConfigResponse>,
+    ) {
+        ctx.spawn(async move {
+            loop {
+                let mut change = GlobalConfigItem::new();
+                change.set_name("/global/config/name".to_owned());
+                change.set_value("changed".to_owned());
```
Comment (zhongzc) [inline components/test_pd/src/server.rs:208]: send two different values to see if the client can receive them.

**Review by zhongzc — APPROVED** [review]

**Pull request merged**